### PR TITLE
Save draw

### PR DESCRIPTION
**Pull request by oterral**
Save the KML automatically in s3.

[TEST](https://mf-geoadmin3.dev.bgdi.ch/teo_save_draw/?lang=en&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&X=190000.00&Y=660000.00&zoom=1&layers=KML%7C%7Chttps:%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fmom_kml_storage%2Ffiles%2FoA-_vdSKS129xxTSMD79Jg)

Translations only in EN

### TODO (from https://github.com/geoadmin/mf-geoadmin3/issues/2315)
- [x] Translations

### CODE REVIEW

**Comment by ponceta** [comment]
Yeah! This is just awesome!

**Comment by cedricmoullet** [comment]
Great !
I think that we should, before the user draws something, explicitely state that the drawing will be stored and provide a link to the terms of use

**Comment by procrastinatio** [comment]
Link with fileID (a green square around Bern)
https://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/shorten/645b10a7ae

You may edit the feature, but the original link remain untouched.

**Comment by procrastinatio** [comment]
jenkins test

**Comment by cedricmoullet** [comment]
work fine !
https://mf-geoadmin3.dev.bgdi.ch/teo_save_draw/src/?lang=en&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&X=190000.00&Y=660000.00&zoom=1&layers=KML%7C%7Chttps:%2F%2Fpublic.dev.bgdi.ch%2FSWKVQsl4RyKuqsO0zl58iA

2015-05-07 20:53 GMT+02:00 procrastinatio <notifications@github.com>:

> jenkins test
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2323#issuecomment-99978061>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://twitter.com/cedricmoullet>

**Comment by cedricmoullet** [comment]
But still an issue with the short link:
https://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/shorten/645b1545df It doesn't
remove the adminId

2015-05-07 20:55 GMT+02:00 Cédric MOULLET <cedric.moullet@gmail.com>:

> work fine !
> https://mf-geoadmin3.dev.bgdi.ch/teo_save_draw/src/?lang=en&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&X=190000.00&Y=660000.00&zoom=1&layers=KML%7C%7Chttps:%2F%2Fpublic.dev.bgdi.ch%2FSWKVQsl4RyKuqsO0zl58iA
>
> 2015-05-07 20:53 GMT+02:00 procrastinatio <notifications@github.com>:
>
>> jenkins test
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/geoadmin/mf-geoadmin3/pull/2323#issuecomment-99978061>
>> .
>>
>
>
>
> --
> Twitter: http://twitter.com/cedricmoullet
> Linked In: http://www.linkedin.com/in/cedricmoullet
>
> <http://twitter.com/cedricmoullet>
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://twitter.com/cedricmoullet>

**Comment by procrastinatio** [comment]
Deleting a single feature does not trigger a POST (update) request, while other _edit_ operations do it correctly.

**Comment by procrastinatio** [comment]
A 2MB KML file is way too big to be efficiently saved [big KML](https://mf-geoadmin3.dev.bgdi.ch/teo_save_draw/?layers=KML%7C%7Chttps:%2F%2Fpublic.dev.bgdi.ch%2F7p5MtZemRY-Oc_VtIiLPaw)

File: src/components/ExportKmlService.js [hunk 1]
```diff
@@ -56,9 +56,9 @@
             clone.getGeometry().transform(projection, 'EPSG:4326');
             var styles;
             if (clone.getStyleFunction()) {
-              styles = clone.getStyleFunction()();
+              styles = clone.getStyleFunction().call(clone);
```
Comment (gjn) [inline src/components/ExportKmlService.js]: Just curious: why was this done?

File: src/locales/fr.json [hunk 2]
```diff
@@ -241,6 +240,9 @@
 	"settings": "Réglages",
 	"share": "Partager",
 	"share_disclaimer": "En intégrant cette carte à votre site, vous acceptez les <a href='http://www.geo.admin.ch/internet/geoportal/fr/home/geoadmin/contact.html#iFrame' target='_blank'> Conditions d'utilisation</a>.",
+	"share_file_disclaimer": "Votre dessin est stocké une année sur une URL publique. En utilisant ce service, vous acceptez les <a href='http://www.geo.admin.ch/internet/geoportal/fr/home/geoadmin/contact.html#drawing ' target='_blank'> Conditions d'utilisation</a>.",
+	"share_file_link_descr_admin": "to fill",
```
Comment (gjn) [inline src/locales/fr.json]: Missing french translation. @cedricmoullet could you provide it?

File: src/locales/it.json [hunk 3]
```diff
@@ -241,6 +240,9 @@
 	"settings": "Impostazioni",
 	"share": "Condividere",
 	"share_disclaimer": "Incorporando questa mappa, accetti i  <a href='http://www.geo.admin.ch/internet/geoportal/fr/home/geoadmin/contact.html#iFrame' target='_blank'> termini di servizio</a>.",
+	"share_file_disclaimer": "Incorporando questa mappa, accetti i  <a href='http://www.geo.admin.ch/internet/geoportal/it/home/geoadmin/contact.html#drawing ' target='_blank'> termini di servizio</a>.",
+	"share_file_link_descr_admin": "to download, change or delete your drawing.",
```
Comment (gjn) [inline src/locales/it.json]: Missing italian translation. @AFoletti ?

File: src/locales/it.json [hunk 4]
```diff
@@ -241,6 +240,9 @@
...
+	"share_file_link_title_admin": "Link for later editing (this link is an administrator link and has not to be shared with others)",
```
Comment (gjn) [inline src/locales/it.json]: @AFoletti

File: src/locales/rm.json [hunk 5]
```diff
@@ -241,6 +240,9 @@
 	"settings": "Configuraziun ",
 	"share": "Parter",
 	"share_disclaimer": "Durch das Einbetten dieser Karte stimmen Sie den  <a href='http://www.geo.admin.ch/internet/geoportal/rm/home/geoadmin/contact.html#iFrame' target='_blank'>Nutzungsbedingungen</a> zu.",
+	"share_file_disclaimer": "Durch das Einbetten dieser Karte stimmen Sie den  <a href='http://www.geo.admin.ch/internet/geoportal/rm/home/geoadmin/contact.html#drawing' target='_blank'>Nutzungsbedingungen</a> zu.",
+	"share_file_link_descr_admin": "to download, change or delete your drawing.",
```
Comment (gjn) [inline src/locales/rm.json]: Should be german...line below as well.

File: src/locales/de.json [hunk 6]
```diff
@@ -241,6 +240,9 @@
 	"settings": "Konfiguration",
 	"share": "Share",
 	"share_disclaimer": "Durch das Einbetten dieser Karte stimmen Sie den  <a href='http://www.geo.admin.ch/internet/geoportal/de/home/geoadmin/contact.html#iFrame' target='_blank'>Nutzungsbedingungen</a> zu.",
+	"share_file_disclaimer": "Durch das Einbetten dieser Karte stimmen Sie den  <a href='http://www.geo.admin.ch/internet/geoportal/de/home/geoadmin/contact.html#drawing ' target='_blank'>Nutzungsbedingungen</a> zu.",
+	"share_file_link_descr_admin": "to download, change or delete your drawing.",
+	"share_file_link_title_admin": "Link for later editing (this link is an administrator link and has not to be shared with others)",
```
Comment (gjn) [inline src/locales/de.json]: Missing german translation

File: test/specs/Loader.spec.js [hunk 7]
```diff
@@ -33,4 +52,37 @@ beforeEach(function() {
...
+  module(function(gaQueryProvider, gaGlobalOptions) {
+    gaQueryProvider.dpUrl = gaGlobalOptions.resourceUrl +
+        'lib/bootstrap-datetimepicker.min.js';
+  });
+
+  module(function($sceDelegateProvider, gaGlobalOptions) {
+    var whitelist = $sceDelegateProvider.resourceUrlWhitelist();
+    whitelist = whitelist.concat(gaGlobalOptions.whitelist);
+    $sceDelegateProvider.resourceUrlWhitelist(whitelist);
+  });
```
Comment (gjn) [inline test/specs/Loader.spec.js]: Heads up for these tests. Well done!

**Comment by gjn** [comment]
This is well done. Added a couple of minor comments.

Translations don't seem to be complete.

I'm not sure about the error handling. All possible errors are simply ignored and the user won't know that a) an error occured and b) what it would mean for his drawing. I agree that error is unlikely, but silent treatement is not optimal.

**Comment by gjn** [comment]
I've rebased on the testlink is up-to-date. We need to fix the issues in https://github.com/geoadmin/mf-geoadmin3/issues/2315

**Comment by gjn** [comment]
jenkins test

**Comment by gjn** [comment]
This is done. Please test again [TEST](https://mf-geoadmin3.dev.bgdi.ch/teo_save_draw/?lang=en&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&X=190000.00&Y=660000.00&zoom=1&layers=KML%7C%7Chttps:%2F%2Fmf-chsdi3.dev.bgdi.ch%2Fmom_kml_storage%2Ffiles%2FoA-_vdSKS129xxTSMD79Jg)

**Comment by procrastinatio** [comment]
I though we add to remove the *one year* from translation (_share_file_link_title_admin_ and _share_file_disclaimer_)  and the text has to be *red*... If now, correct the google doc and ignore my commit #c9396cc

**Comment by gjn** [comment]
We just went throuhg all this...see https://github.com/geoadmin/mf-geoadmin3/issues/2315

Please let them test and  the correct when it's wrong. I'll remove the commit.

**Comment by gjn** [comment]
Removed the superflous `:` (good find). Testlink up to date.

**Comment by gjn** [comment]
jenkins test

**Comment by gjn** [comment]
jenkins test

**Comment by gjn** [comment]
I've rebased. Now the test link is including the third party warning stuff. So you can test this together. I've also fixed the code that broke the print button text (found thanks to jenkins).

This can be merged once people are happy @davidoesch @cedricmoullet .

**Comment by davidoesch** [comment]
me happy

**Comment by cedricmoullet** [comment]
Excellent job ! Thanks to all making this possible.

2015-05-21 17:02 GMT+02:00 Gilbert Jeiziner <notifications@github.com>:

> Merged #2323 <https://github.com/geoadmin/mf-geoadmin3/pull/2323>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2323#event-310745768>.
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://twitter.com/cedricmoullet>

**Pull request merged**